Mathlib/MeasureTheory/Function/JacobianOneDim.lean

```diff
@@ -60,6 +66,319 @@ theorem integral_image_eq_integral_abs_deriv_smul
+      simp only [mem_inter_iff, mem_diff, hx.1.1, mem_union, mem_setOf_eq, true_and, not_or,
+        mem_preimage, b, s₁, a] at hx
+      exact neBot_iff.2 hx.1.1
+  · /- We have to show that the derivative is nonnegative at points of `c`. As these points are
+    not isolated in `s`, this follows from the fact that `f` is monotone on `s`. -/
+    intro x hx
+    apply (hf' x hx.1.1).nonneg_of_monotoneOn _ hf
+    simp only [mem_diff, hx.1.1, mem_union, mem_setOf_eq, true_and, not_or, c, s₁, a, b] at hx
+    rw [accPt_principal_iff_nhdsWithin]
+    have : (𝓝[s ∩ Iio x] x).NeBot := neBot_iff.2 hx.1.2
+    apply this.mono
+    apply nhdsWithin_mono
+    rintro y ⟨yt, yx : y < x⟩
+    exact ⟨yt, by simpa using yx.ne⟩
+  · intro x hx y hy hxy
+    contrapose! hxy
+    wlog H : x < y generalizing x y with h
+    · have : y < x := by order
+      exact (h hy hx hxy.symm this).symm
+    refine fun h ↦ hx.2 ⟨hx.1, ?_⟩
+    exact ⟨x, y, hx.1, hy.1, H, rfl, h.symm⟩
+
+/- Change of variable formula for differentiable functions: if a real function `f` is
+monotone and differentiable on a measurable set `s`, then the Lebesgue integral of a function
+`u : ℝ → ℝ≥0∞` on `f '' s` coincides with the integral of `(f' x) * u ∘ f` on `s`.
+Note that the measurability of `f '' s` is given by `MeasurableSet.image_of_monotoneOn`. -/
+theorem lintegral_image_eq_lintegral_deriv_mul_of_monotoneOn (hs : MeasurableSet s)
+    (hf' : ∀ x ∈ s, HasDerivWithinAt f (f' x) s x) (hf : MonotoneOn f s) (u : ℝ → ℝ≥0∞) :
+    ∫⁻ x in f '' s, u x = ∫⁻ x in s, ENNReal.ofReal (f' x) * u (f x) := by
+  rcases exists_decomposition_of_monotoneOn_hasDerivWithinAt hs hf hf' with
+    ⟨a, b, c, h_union, ha, hb, hc, h_disj, h_disj', a_count, fb_count, deriv_b, deriv_c, inj_c⟩
+  have I : ∫⁻ x in s, ENNReal.ofReal (f' x) * u (f x)
+      = ∫⁻ x in c, ENNReal.ofReal (f' x) * u (f x) := by
+    have : ∫⁻ x in a, ENNReal.ofReal (f' x) * u (f x) = 0 := by
+      have : volume a = 0 := a_count.measure_zero volume
+      exact setLIntegral_measure_zero a _ this
+    rw [← h_union, lintegral_union (hb.union hc) h_disj, this, zero_add]
+    have : ∫⁻ x in b, ENNReal.ofReal (f' x) * u (f x) = 0 :=
+      setLIntegral_eq_zero hb (fun x hx ↦ by simp [deriv_b x hx])
+    rw [lintegral_union hc h_disj', this, zero_add]
+  have J : ∫⁻ x in f '' s, u x = ∫⁻ x in f '' c, u x := by
+    apply setLIntegral_congr
+    rw [← h_union, image_union, image_union]
+    have A : (f '' a ∪ (f '' b ∪ f '' c) : Set ℝ) =ᵐ[volume] (f '' b ∪ f '' c : Set ℝ) := by
+      refine union_ae_eq_right_of_ae_eq_empty (ae_eq_empty.mpr ?_)
+      exact (a_count.image _).measure_zero _
+    have B : (f '' b ∪ f '' c : Set ℝ) =ᵐ[volume] f '' c := by
+      refine union_ae_eq_right_of_ae_eq_empty (ae_eq_empty.mpr ?_)
+      exact fb_count.measure_zero _
+    exact A.trans B
+  rw [I, J]
+  have c_s : c ⊆ s := by rw [← h_union]; exact subset_union_right.trans subset_union_right
```

Comment:
```suggestion
  have c_s : c ⊆ s := h_union ▸ subset_union_right.trans subset_union_right

Reply:
With my heuristic of " ▸ is a simpler form of `rw`" (but applying in either direction), I don't find it that bad.
On the contrary, all the `by rw ... exact` is lots of space for not so much information. But anyway, you're free to disagree here!